keps/sig-instrumentation/draft-20190404-kubernetes-control-plane-metrics-stability.md

```diff
@@ -0,0 +1,248 @@
+
+ 1. _soft deprecated_ - this occurs when a stable metric has been recently deprecated; 'recently deprecated' is defined as when the metric's deprecated version is equal to the current kubernetes version
+ 2. _hard deprecated_ - this occurs when the metric has been deprecated for > 1 release.
+
+Soft deprecated metrics will have their description text prefixed with a deprecation notice string '(Deprecated from x.y)' and a warning log will be emitted during metric registration (in the spirit of the official [kubernetes deprecation policy](https://kubernetes.io/docs/reference/using-api/deprecation-policy/#deprecating-a-flag-or-cli)). Like their stable metric counterparts, soft deprecated metrics will be automatically registered to the metrics endpoint.
+
+Metrics are _hard deprecated_ once they have been deprecated for more than one release. _Unlike_ their soft deprecated counterparts, hard deprecated metrics will __*no longer be automatically registered*__ to the metrics endpoint. By default, hard deprecated metrics will not be registered unless explicitly instructed to do so through command line flag on the binary (i.e. '--enable-hard-deprecated-metrics=all-metrics'). It is expected that deprecated metrics will be removed once they have been deprecated for more than one release. As such, metrics which have been deprecated for more than one release will have __no__ stability guarantees.
+
+## Design Details
+
+### Test Plan
+
+Internal wrappers can be unit tested. There has been some discussion around providing APIs around metric definitions, hopefully with the end goal of being able to augment our test strategy with current and historical metric definition data.
+
+### Graduation Criteria
+
+TBD
+
+## Drawbacks
+
+More generally, this proposal has the drawbacks which any proposal suggesting a more rigorous enforcement of an API is going to have. There is always a tradeoff between the ease at which a developer can make breaking changes to an API, with consumers' ability to reliably use that API.
+
+Relative to a more hands-off approach, like one where we just document the metrics which the community has agreed to 'certify' as stable, this approach is definitely more heavyweight. This approach involves more code and more code is more maintenance. However, most of the code will be centralized and the internal logic is easily unit-testable. We also do not have to worry much about changing internal API semantics, since our wrappers will be used internally only, which means it should be easy to modify for new usecases in the future. This sort of approach also enables static analysis tooling around metrics which we could run in precommit.
+
+Also, we should note that this approach can be manufactured in-place; this framework could be rolled out without actually introducing any backwards-incompatible changes (unlike moving stable metrics to a '/metrics/v1' endpoint).
+
+
```

Comment:
I think you need to discuss a here a bit what happens if a metric is no longer relevant (e.g. it's code is re-architected in such a way that it's no longer feasible to provide that metric, there's no longer anything to measure, etc).  We can certainly provide guidance on "don't make a metric stable if there's any way for it to get refactored away", but a plan for what happens if we go to break that (do we block the feature?  go "oops, oh well?") is probably a good idea.

Reply:
Yes, generally I think that guidance is reasonable (i.e. 'don't make a metric stable if there's any way for it to get refactored away') but in the case where it does get refactored away, we should immediately deprecate the metric and go through the normal deprecation/removal policy. The intent of this stability proposal is to safeguard the consumption pipeline. 

For better or worse, I have been considering out-of-scope any breakages which are caused by alerts which make assumptions based specific values in metrics. I should probably be more explicit about that though.

Reply:
yeah, just saying that would be a good start

Reply:
Added an explicit non-goal to the document (i.e. addressing breakages in alerting based of off assumptions on specific values in metrics) and made it more explicit that the stability contract specifically targets metric consumption rather than anything derived from specific metric data. Also added a more targeted paragraph in the drawbacks section regarding this specific point.